### DISCUSSION TITLE
How to open multiple different local web apps in one tauri-app?

### DISCUSSION THREAD

**Discussion by XX** [tauri-apps/tauri#11861] [Q&A]
I'm making something similar to a superapp. I have a main web app and a bunch of other web apps installed into it. The `dist` structure is something like this:
```
dist
  wapp_a
    index.html
    wapp_a.js
    wapp_a.wasm
  wapp_b
    index.html
    wapp_b.js
    wapp_b.wasm
  ...
  index.html
  main.js
  main.wasm
```
The main application is a manager for other applications (`wapp_a`, `wapp_b`, ...) and can launch them. I can create a separate window by calling the command that creates `WebviewWindowBuilder`, but how do I specify that the new WebviewWindow should open `dist/wapp_a/index.html`, and not `dist/index.html`? If I specify a qualifying path in `WebviewUrl::App("wapp_a/index.html".into())`, it still opens the main `index.html`, although in the network tab in devtools I see a call to `tauri://localhost/wapp_a/index.html`.

How can I do this? And if it is possible, how can I isolate the new WebviewWindow so that `wapp_a` can't access `wapp_b` assets? If it is not possible, where in the code are there any modifications I could make to make this possible? If this is not possible, where in the Tauri code are there improvements that I could make to make this possible?

**Comment by FabianLars** [comment]
> If I specify a qualifying path in WebviewUrl::App("wapp_a/index.html".into()), it still opens the main index.html, although in the network tab in devtools I see a call to tauri://localhost/wapp_a/index.html.

hmm, that's supposed to work. A long time ago i did the same too.

> how can I isolate the new WebviewWindow so that wapp_a can't access wapp_b assets?

that's indeed not supported. can't really point to the best entry point in the code without knowing how to best solve this but this file is where the "server" is https://github.com/tauri-apps/tauri/blob/dev/crates/tauri/src/protocol/tauri.rs

>

**Comment by XX** [comment]
@FabianLars 
> hmm, that's supposed to work. A long time ago i did the same too.

Maybe I'm doing something wrong? But in my example the second app always opens the main one (It's interesting that the links to static images are broken):

![Screenshot_20241204_151645](https://github.com/user-attachments/assets/278023a8-4f9b-4585-924b-83241abee56e)

The source code here: https://github.com/XX/tauri-superapp

**Reply by FabianLars** [reply]
In the reproduction at least the frontend-dist is wrong. If you point it to dist/main/ it will only include dist/main/ but _not_ dist/wapp_a which then makes tauri fall back to main for the second window.

**Reply by XX** [reply]
Yes, thank you! This solves half the problem:

![Screenshot_20241204_165735](https://github.com/user-attachments/assets/0b2becfc-5331-4fda-9533-310575747a4f)

But the web apps do not have their own root directory, they are as if parts of one main web app, the paths must be specified from the root `dist`:

```js
import init from '/wapp_a/wapp_a.js';
```

**Reply by XX** [reply]
Maybe it's possible to set up a hook where I can intercept any access to assets and prevent the sub-application from going beyond its directory?

**Reply by FabianLars** [reply]
> But the web apps do not have their own root directory, they are as if parts of one main web app, the paths must be specified from the root dist:

Right, that sounds likely

> Maybe it's possible to set up a hook where I can intercept any access to assets and prevent the sub-application from going beyond its directory?

https://docs.rs/tauri/2.0.0/tauri/webview/struct.WebviewWindowBuilder.html#method.on_navigation and https://docs.rs/tauri/2.0.0/tauri/webview/struct.WebviewWindowBuilder.html#method.on_web_resource_request maybe. Alternatively you could always write your own "server" via https://docs.rs/tauri/2.0.0/tauri/struct.Builder.html#method.register_asynchronous_uri_scheme_protocol (or its sync variant) - that's basically the api tauri uses for the default "server".

**Reply by XX** [reply]
@FabianLars Thanks! I'll try to do this.

**Reply by XX** [reply]
Hmm... I logged the URL in `on_navigation` and `on_web_resource_request` interceptors but only saw the `tauri://localhost/wapp_a/index.html` in the output. On the Network tab in devtools I see an additional `tauri://localhost/wapp_a/wapp_a.js` line, but it's not in the log..

**Reply by XX** [reply]
Similar issue: https://github.com/tauri-apps/tauri/issues/4899

**Reply by XX** [reply]
@FabianLars Is there a way to intercept all requests via the `tauri://` protocol? (For `wapp_a` - `tauri://localhost/wapp_a/index.html` and `tauri://localhost/wapp_a/wapp_a.js` in my case.)

**Reply by FabianLars** [reply]
hmm in my app on macos at least on_web_resource_request seems to include all files 🤔

**Reply by XX** [reply]
On my Kubuntu linux:

![Screenshot_20241204_193020_x](https://github.com/user-attachments/assets/01652adb-43bd-416a-bbc6-3243bad00251)
